pipeline.libsonnet

```diff
@@ -0,0 +1,187 @@
+    local file_suffix = std.strReplace(tag, '-', '.');
+    local volumes = if is_windows then [{ name: windows_pipe_volume, path: windows_pipe }] else [];
+    local golang = golang_image(os, version);
+    local plugin_repo = 'plugins/' + std.split(name, '-')[1];
+    local extension = if is_windows then '.exe' else '';
+    {
+      kind: 'pipeline',
+      name: tag,
+      platform: {
+        os: os,
+        arch: arch,
+        version: if std.length(version) > 0 then version,
+      },
+      steps: [
+        {
+          name: 'build-push',
+          image: golang,
+          pull: 'always',
+          environment: {
+            CGO_ENABLED: '0',
+            GO111MODULE: 'on',
+          },
+          commands: [
+            'go build -v -ldflags "-X main.build=${DRONE_BUILD_NUMBER}" -a -o release/' + os + '/' + arch + '/' + name + extension,
+          ],
+          when: {
+            event: ['push', 'pull_request'],
+          },
+        },
+        {
+          name: 'build-tag',
+          image: golang,
+          pull: 'always',
+          environment: {
+            CGO_ENABLED: '0',
+            GO111MODULE: 'on',
+          },
+          commands: [
+            'go build -v -ldflags "-X main.version=${DRONE_TAG##v} -X main.build=${DRONE_BUILD_NUMBER}" -a -o release/' + os + '/' + arch + '/' + name + extension,
+          ],
+          when: {
+            event: ['tag'],
+          },
+        },
+        {
+          name: 'executable',
+          image: golang,
+          pull: 'always',
+          commands: [
+            './release/' + os + '/' + arch + '/' + name + extension + ' --help',
+          ],
+        },
+        {
+          name: 'dryrun',
+          image: 'plugins/docker:' + tag,
+          pull: 'always',
+          settings: {
+            dry_run: true,
+            tags: tag,
+            dockerfile: 'docker/Dockerfile.' + file_suffix,
+            daemon_off: if is_windows then 'true' else 'false',
+            repo: plugin_repo,
+            username: { from_secret: 'docker_username' },
+            password: { from_secret: 'docker_password' },
+          },
+          volumes: if std.length(volumes) > 0 then volumes,
+          when: {
+            event: ['pull_request'],
+          },
+        },
+        {
+          name: 'publish',
+          image: 'plugins/docker:' + tag,
+          pull: 'always',
+          settings: {
+            auto_tag: true,
+            auto_tag_suffix: tag,
+            daemon_off: if is_windows then 'true' else 'false',
+            dockerfile: 'docker/Dockerfile.' + file_suffix,
+            repo: plugin_repo,
+            username: { from_secret: 'docker_username' },
+            password: { from_secret: 'docker_password' },
+          },
+          volumes: if std.length(volumes) > 0 then volumes,
+          when: {
+            event: ['push'],
+          },
+        },
+      ],
+      trigger: {
```

Comment:
Currently this won't trigger builds for tags, but @bradrydzewski wants to add `ref` as a trigger, than we could do something like this:

```
      trigger: {
        ref: ['refs/heads/master', 'refs/pulls/**', 'refs/tags/**'],
      },
```